# Challenge 3

In this challenge we will work on the `Orders` data set. In your work you will apply the thinking process and workflow we showed you in Challenge 2.

You are serving as a Business Intelligence Analyst at the headquarter of an international fashion goods chain store. Your boss today asked you to do two things for her:

**First, identify two groups of customers from the data set.** The first group is **VIP Customers** whose **aggregated expenses** at your global chain stores are **above the 95th percentile** (aka. 0.95 quantile). The second group is **Preferred Customers** whose **aggregated expenses** are **between the 75th and 95th percentile**.

**Second, identify which country has the most of your VIP customers, and which country has the most of your VIP+Preferred Customers combined.**

## Q1: How to identify VIP & Preferred Customers?

We start by importing all the required libraries:

In [1]:
# import required libraries
import pandas as pd
import numpy as np
import sqlalchemy
import datetime

Next, import `Orders` from Ironhack's database into a dataframe variable called `orders`. Print the head of `orders` to overview the data:

**Solution:** Import the data from the database. Nothing different from the other parts of the lab apart from the name of the database.

In [2]:
# your code here
# import data set from Ironhack's database
driver = 'mysql+pymysql:'
user = 'USERNAME'
password = 'PASSWORD'
ip = 'IP'
database = 'orders'

engine = sqlalchemy.create_engine(connection_string)
engine.table_names()

['orders']

In [3]:
query = 'SELECT * FROM orders;'
orders = pd.read_sql(query, engine)

**Note:** It's always good practice to take an initial look at the data and see if it needs some cleaning and adjustments. For example we can check if there any NaN values and see if the types of the columns make sense.

In [14]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 397924 entries, 0 to 397923
Data columns (total 14 columns):
index           397924 non-null int64
InvoiceNo       397924 non-null int64
StockCode       397924 non-null object
year            397924 non-null int64
month           397924 non-null int64
day             397924 non-null int64
hour            397924 non-null int64
Description     397924 non-null object
Quantity        397924 non-null int64
InvoiceDate     397924 non-null object
UnitPrice       397924 non-null float64
CustomerID      397924 non-null int64
Country         397924 non-null object
amount_spent    397924 non-null float64
dtypes: float64(2), int64(8), object(4)
memory usage: 45.5+ MB


In [15]:
orders.head()

index  InvoiceNo StockCode  year  month  day  hour  \
0      0     536365    85123A  2010     12    3     8   
1      1     536365     71053  2010     12    3     8   
2      2     536365    84406B  2010     12    3     8   
3      3     536365    84029G  2010     12    3     8   
4      4     536365    84029E  2010     12    3     8   

                           Description  Quantity          InvoiceDate  \
0   white hanging heart t-light holder         6  2010-12-01 08:26:00   
1                  white metal lantern         6  2010-12-01 08:26:00   
2       cream cupid hearts coat hanger         8  2010-12-01 08:26:00   
3  knitted union flag hot water bottle         6  2010-12-01 08:26:00   
4       red woolly hottie white heart.         6  2010-12-01 08:26:00   

   UnitPrice  CustomerID         Country  amount_spent  
0       2.55       17850  United Kingdom         15.30  
1       3.39       17850  United Kingdom         20.34  
2       2.75       17850  United Kingdom         22.00  
3       3.39       17850  United Kingdom         20.34  
4       3.39       17850  United Kingdom         20.34

As we can see there are no NaN values, since every column has the same number of non-null values as the number of entries: 397924.

The InvoiceDate columns is currently of `object` type, which means that it's a string. We can convert it to datetime so it'll be easier to work with.

Also the `index` column seems kind of redundant since we already have the pandas index. We will drop it.

In [17]:
orders['InvoiceDate'] = pd.to_datetime(orders['InvoiceDate'])

In [19]:
orders.drop('index', axis=1, inplace=True)

---

"Identify VIP and Preferred Customers" is the non-technical goal of your boss. You need to translate that goal into technical languages that data analysts use:

## How to label customers whose aggregated `amount_spent` is in a given quantile range?


We break down the main problem into several sub problems:

#### Sub Problem 1: How to aggregate the  `amount_spent` for unique customers?

#### Sub Problem 2: How to select customers whose aggregated `amount_spent` is in a given quantile range?

#### Sub Problem 3: How to label selected customers as "VIP" or "Preferred"?

*Note: If you want to break down the main problem in a different way, please feel free to revise the sub problems above.*

Now in the workspace below, tackle each of the sub problems using the iterative problem solving workflow. Insert cells as necessary to write your codes and explain your steps.

**Solution Sub Problem 1:**  
First of all we need to aggregate the amount_spent for each customer. This can be easily done by grouping the dataframe by `CustomerID` and then using the `sum()` function to aggregate the values.

In [24]:
customers_agg = orders.groupby('CustomerID').sum()
customers_agg.head()

InvoiceNo    year  month  day  hour  Quantity  UnitPrice  \
CustomerID                                                             
12346          541431    2011      1    2    10     74215       1.04   
12347       101296926  365971   1383  441  2219      2458     481.21   
12348        16869685   62324    257  111   472      2341     178.71   
12349        42165457  146803    803   73   657       631     605.10   
12350         9231629   34187     34   51   272       197      65.30   

            amount_spent  
CustomerID                
12346           77183.60  
12347            4310.00  
12348            1797.24  
12349            1757.55  
12350             334.40

Note that we only need the sum of the amount_spent, the other columns are not giving us any useful information since we're just summing Invoice numbers, years, months, etc.

We will therefore drop everything except `amount_spent`. We are using double `[[]]` around the column name so that we will save a DataFrame and not a Series. The only actual difference here is in the way in which a Jupyter notebook prints a DataFrame and a Series: it simply looks nicer in the first case.

In [25]:
customers_agg = customers_agg[['amount_spent']]

In [27]:
customers_agg.head()

amount_spent
CustomerID              
12346           77183.60
12347            4310.00
12348            1797.24
12349            1757.55
12350             334.40

**Solution Sub Problem 2:**  
We now need to select which customers fall in the quantile ranges that we were given at the start of the exercise, that is 95% and 75%. We can use the `quantile()` method to see where the two quantile falls. Anything higher than the breakpoint is part of that specific quantile. We then select them by filtering the database.

In [33]:
customers_agg.quantile([0.75, 0.95])

amount_spent
0.75      1661.640
0.95      5840.182

In [43]:
q_75 = customers_agg.quantile(0.75)[0]
q_95 = customers_agg.quantile(0.95)[0]

In [44]:
q_75, q_95

(1661.6400000000003, 5840.181999999983)

In [48]:
customers_75 = customers_agg.loc[customers_agg['amount_spent'] > q_75]

In [49]:
customers_95 = customers_agg.loc[customers_agg['amount_spent'] > q_95]

**Solution Sub Problem 3:**  
Now that we know which customers are in the 75 and 95 percentile by amount_spent we can simply use that information to apply the relevant labels, that is VIP and Preferred. We can therefore create a new column in the original aggregate DataFrame that will have `VIP` for the 95 percentile and `Preferred` for the 75. There are a lot of ways to do this, but just for the sake of looking at a new function we can use the numpy `select` function, which needs a list of conditions paired with a list of outputs. The doc is here: https://docs.scipy.org/doc/numpy/reference/generated/numpy.select.html

In [58]:
conditions = [customers_agg['amount_spent'] > q_95, customers_agg['amount_spent'] > q_75] 
outputs = ['VIP', 'Preferred']

customers_agg['label'] = np.select(conditions, outputs, default=None)

In [63]:
customers_agg.head()

amount_spent      label
CustomerID                         
12346           77183.60        VIP
12347            4310.00  Preferred
12348            1797.24  Preferred
12349            1757.55  Preferred
12350             334.40       None

## Q2: How to identify which country has the most VIP Customers?

**Solution:**  
Now that we have labelled the customers in the `customers_agg` dataframe, which also contains the relevant `CustomerID`, we can simply extract the related information and apply it to the original `orders` dataframe. Then we can group the dataframe by country and see which one has the most by aggregating the data using the `count` method.

One of the fastest way is joining the two dataframes on the `CustomerID` column so that the labels will be automatically applied to the relevant CustomerIDs.

In [69]:
# your code here
orders = orders.join(customers_agg['label'], on='CustomerID')

In [75]:
orders_vip = orders.loc[orders['label'] == 'VIP']

In [77]:
orders_vip.groupby('Country').count()['CustomerID'].sort_values(ascending=False).head()

Country
United Kingdom    84185
EIRE               7077
France             3290
Germany            3127
Netherlands        2080
Name: CustomerID, dtype: int64

As we can see the country with the most VIP customers is the United Kingdom with 84185.

## Q3: How to identify which country has the most VIP+Preferred Customers combined?

**Solutions:**  
The solution is the same as Q2, we simply take all the rows where the label is either VIP or Preferred.

In [79]:
# your code here
orders_vip_pref = orders.loc[(orders['label'] == 'VIP') | (orders['label'] == 'Preferred')]

In [81]:
orders_vip_pref.groupby('Country').count()['CustomerID'].sort_values(ascending=False).head()

Country
United Kingdom    221635
Germany             7349
EIRE                7238
France              6301
Netherlands         2080
Name: CustomerID, dtype: int64

The United Kingdom is still the country with most customers when only taking into account VIP and Preferred ones.